File: app/src/main/java/com/akhbulatov/wordkeeper/presentation/ui/words/WordsFragment.java
package com.akhbulatov.wordkeeper.presentation.ui.words;

import android.app.SearchManager;
import android.content.ComponentName;
import android.content.Context;
import android.os.Bundle;
import android.text.Html;
import android.text.TextUtils;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.EditorInfo;

import com.akhbulatov.wordkeeper.App;
import com.akhbulatov.wordkeeper.R;
import com.akhbulatov.wordkeeper.databinding.FragmentWordsBinding;
import com.akhbulatov.wordkeeper.domain.global.models.Word;
import com.akhbulatov.wordkeeper.presentation.ui.addeditword.AddEditWordDialog;
import com.akhbulatov.wordkeeper.presentation.ui.global.base.BaseFragment;
import com.akhbulatov.wordkeeper.presentation.ui.global.list.adapters.WordAdapter;
import com.akhbulatov.wordkeeper.presentation.ui.global.models.WordUiModel;
import com.akhbulatov.wordkeeper.presentation.ui.global.models.WordUiModelKt;
import com.akhbulatov.wordkeeper.presentation.ui.selectwordcategory.SelectWordCategoryDialog;
import com.akhbulatov.wordkeeper.presentation.ui.sortword.SortWordDialog;
import com.akhbulatov.wordkeeper.presentation.ui.main.MainActivity;
import com.akhbulatov.wordkeeper.util.CommonUtils;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.view.ActionMode;
import androidx.appcompat.widget.SearchView;
import androidx.fragment.app.DialogFragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.DividerItemDecoration;

public class WordsFragment extends BaseFragment {

    private WordAdapter mWordAdapter;

    private ActionModeCallback mActionModeCallback;
    private ActionMode mActionMode;

    @Inject ViewModelProvider.Factory viewModelFactory;
    private WordsViewModel viewModel;

    private FragmentWordsBinding binding;

    public WordsFragment() {
        super(R.layout.fragment_words);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        App.appComponent
                .wordsComponentFactory()
                .create()
                .inject(this);
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);

        viewModel = new ViewModelProvider(this, viewModelFactory).get(WordsViewModel.class);
        viewModel.loadWords();

        mWordAdapter = new WordAdapter(new WordAdapter.OnItemClickListener() {
            @Override
            public void onItemClick(@NotNull Word word, int position) {
                if (mActionMode != null) {
                    toggleSelection(position);
                }
            }

            @Override
            public boolean onItemLongClick(@NotNull Word word, int position) {
                if (mActionMode == null) {
                    mActionMode = ((AppCompatActivity) requireActivity()).startSupportActionMode(mActionModeCallback);
                }
                toggleSelection(position);
                return true;
            }
        });

        mActionModeCallback = new ActionModeCallback();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        binding = FragmentWordsBinding.bind(view);
        binding.wordsRecyclerView.setHasFixedSize(true);
        binding.wordsRecyclerView.addItemDecoration(new DividerItemDecoration(requireContext(), DividerItemDecoration.VERTICAL));
        binding.wordsRecyclerView.setAdapter(mWordAdapter);
        binding.addWordFab.setOnClickListener(v -> showAddEditWordDialog(null));

        viewModel.getViewState().observe(getViewLifecycleOwner(), this::renderViewState);
    }

    @Override
    public void onDestroyView() {
        binding.wordsRecyclerView.setAdapter(null);
        binding = null;
        super.onDestroyView();
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        inflater.inflate(R.menu.fragment_word, menu);

        MenuItem searchItem = menu.findItem(R.id.menu_search_word);
        SearchView searchView = (SearchView) searchItem.getActionView();

        SearchManager searchManager = (SearchManager) requireActivity().getSystemService(Context.SEARCH_SERVICE);
        if (searchManager != null) {
            searchView.setSearchableInfo(searchManager
                    .getSearchableInfo(new ComponentName(requireActivity(), MainActivity.class)));
        }
        searchView.setImeOptions(EditorInfo.IME_ACTION_SEARCH);
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
//                final Cursor cursor = mWordDbAdapter.getAll(sSortMode);
//                final int column = cursor.getColumnIndex(WordEntry.COLUMN_NAME);
                if (newText.length() > 0) {
//                    mWordAdapter.swapCursor(new FilterCursorWrapper(cursor, newText, column));

                    binding.emptyDataTextView.setVisibility(View.GONE);
                    if (mWordAdapter.getItemCount() == 0) {
                        String escapedNewText = TextUtils.htmlEncode(newText);
                        String formattedNoResults = String.format(
                                getString(R.string.words_no_results), escapedNewText);
                        CharSequence styledNoResults = Html.fromHtml(formattedNoResults);

                        binding.noResultsTextView.setText(styledNoResults);
                        binding.noResultsTextView.setVisibility(View.VISIBLE);
                    } else {
                        binding.noResultsTextView.setVisibility(View.GONE);
                    }
                } else {
//                    mWordAdapter.swapCursor(cursor);

                    binding.noResultsTextView.setVisibility(View.GONE);
                    if (mWordAdapter.getItemCount() == 0) {
                        binding.noResultsTextView.setVisibility(View.VISIBLE);
                    }
                }
                return true;
            }
        });
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.menu_sort_word) {
            showWordSortDialog();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void renderViewState(WordsViewModel.ViewState viewState) {
        showEmptyProgress(viewState.getEmptyProgress());
        showEmptyData(viewState.getEmptyData());
        showEmptyError(viewState.getEmptyError().getFirst(), viewState.getEmptyError().getSecond());
        showWords(viewState.getWords().getFirst(), viewState.getWords().getSecond());
    }

    private void showEmptyProgress(boolean show) {
        binding.emptyProgressBar.setVisibility(show ? View.VISIBLE : View.GONE);
    }

    private void showEmptyData(boolean show) {
        binding.emptyDataTextView.setVisibility(show ? View.VISIBLE : View.GONE);
    }

    private void showEmptyError(boolean show, String message) {
        binding.errorTextView.setText(message);
        binding.errorTextView.setVisibility(show ? View.VISIBLE : View.GONE);
    }

    private void showWords(boolean show, List<Word> words) {
        mWordAdapter.submitList(words);
        binding.wordsRecyclerView.setVisibility(show ? View.VISIBLE : View.GONE);
    }

    private void showAddEditWordDialog(Word word) {
        WordUiModel uiModel = null;
        if (word != null) {
            uiModel = WordUiModelKt.toUiModel(word);
        }
        AddEditWordDialog.newInstance(uiModel).show(getParentFragmentManager(), null);
    }

    private void toggleSelection(int position) {
        mWordAdapter.toggleSelection(position);
        int count = mWordAdapter.getSelectedWordCount();

        if (count == 0) {
            mActionMode.finish();
        } else {
            mActionMode.setTitle(String.valueOf(count));
            mActionMode.invalidate();
        }
    }

    private void showWordSortDialog() {
        getParentFragmentManager().setFragmentResultListener(
                SortWordDialog.REQUEST_SORT_MODE,
                getViewLifecycleOwner(),
                (requestKey, result) -> viewModel.loadWords()
        );
        DialogFragment dialog = new SortWordDialog();
        dialog.show(requireActivity().getSupportFragmentManager(), null);
    }

    private void showCategoryListDialog() {
        getParentFragmentManager().setFragmentResultListener(
                SelectWordCategoryDialog.REQUEST_SELECT_WORD_CATEGORY,
                getViewLifecycleOwner(),
                (requestKey, result) -> {
                    String category = result.getString(SelectWordCategoryDialog.RESULT_SELECT_WORD_CATEGORY);
                    List<Word> words = new ArrayList<>();
                    for (int pos : mWordAdapter.getSelectedWords()) {
                        words.add(mWordAdapter.getCurrentList().get(pos));
                    }
                    viewModel.onSelectWordCatalogClicked(words, category);
                    mActionMode.finish();
                    CommonUtils.showToast(getActivity(), R.string.words_success_move);
                }
        );
        DialogFragment dialog = new SelectWordCategoryDialog();
        dialog.show(requireActivity().getSupportFragmentManager(), null);
    }

    private class ActionModeCallback implements ActionMode.Callback {

        private MenuItem mItemEditWord;

        @Override
        public boolean onCreateActionMode(ActionMode mode, Menu menu) {
            mode.getMenuInflater().inflate(R.menu.selected_word, menu);
            mItemEditWord = menu.findItem(R.id.menu_edit_word);
            return true;
        }

        @Override
        public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
            // Editing is available only for one selected word
            if (mWordAdapter.getSelectedWordCount() == 1) {
                mItemEditWord.setVisible(true);
            } else {
                mItemEditWord.setVisible(false);
            }
            return true;
        }

        @Override
        public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
            switch (item.getItemId()) {
                case R.id.menu_move_word:
                    showCategoryListDialog();
                    return true;
                case R.id.menu_edit_word:
                    int position = mWordAdapter.getSelectedWords().get(0);
                    Word word = mWordAdapter.getCurrentList().get(position);
                    showAddEditWordDialog(word);
                    mode.finish();
                    return true;
                case R.id.menu_delete_word:
                    List<Word> words = new ArrayList<>();
                    for (int pos : mWordAdapter.getSelectedWords()) {
                        words.add(mWordAdapter.getCurrentList().get(pos));
                    }
                    viewModel.onDeleteWordsClicked(words);
                    mode.finish();
                    return true;
                default:
                    return false;
            }
        }

        @Override
        public void onDestroyActionMode(ActionMode mode) {
            mWordAdapter.clearSelection();
            mActionMode = null;
        }
    }
}

File: app/src/main/java/com/akhbulatov/wordkeeper/presentation/ui/words/WordsFragment.kt
package com.akhbulatov.wordkeeper.presentation.ui.words

import android.app.SearchManager
import android.content.ComponentName
import android.content.Context
import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.inputmethod.EditorInfo
import androidx.appcompat.view.ActionMode
import androidx.appcompat.widget.SearchView
import androidx.core.text.htmlEncode
import androidx.core.text.parseAsHtml
import androidx.core.view.isVisible
import androidx.fragment.app.setFragmentResultListener
import androidx.fragment.app.viewModels
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.RecyclerView
import com.akhbulatov.wordkeeper.R
import com.akhbulatov.wordkeeper.core.ui.base.BaseFragment
import com.akhbulatov.wordkeeper.core.ui.list.adapters.WordAdapter
import com.akhbulatov.wordkeeper.core.ui.models.toUiModel
import com.akhbulatov.wordkeeper.core.ui.utils.requireCompatActivity
import com.akhbulatov.wordkeeper.core.ui.utils.showSnackbar
import com.akhbulatov.wordkeeper.databinding.FragmentWordsBinding
import com.akhbulatov.wordkeeper.domain.word.models.Word
import com.akhbulatov.wordkeeper.presentation.ui.addeditword.AddEditWordDialog
import com.akhbulatov.wordkeeper.presentation.ui.main.MainActivity
import com.akhbulatov.wordkeeper.presentation.ui.selectwordcategory.SelectWordCategoryDialog
import com.akhbulatov.wordkeeper.presentation.ui.sortwords.SortWordsDialog
import javax.inject.Inject

class WordsFragment : BaseFragment(R.layout.fragment_words) {

    @Inject lateinit var viewModelFactory: ViewModelProvider.Factory
    private val viewModel by viewModels<WordsViewModel> { viewModelFactory }

    private var _binding: FragmentWordsBinding? = null
    private val binding get() = _binding!!

    private val wordActionModeCallback by lazy { WordActionModeCallback() }
    private var wordActionMode: ActionMode? = null

    private val wordAdapter by lazy {
        WordAdapter(object : WordAdapter.OnItemClickListener {
            override fun onItemClick(position: Int) {
                if (wordActionMode != null) {
                    toggleSelection(position)
                }
            }

            override fun onItemLongClick(position: Int): Boolean {
                if (wordActionMode == null) {
                    wordActionMode = requireCompatActivity().startSupportActionMode(wordActionModeCallback)
                }
                toggleSelection(position)
                return true
            }
        })
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        WordsComponent.create().inject(this)
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
        if (savedInstanceState == null) {
            viewModel.loadWords()
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        _binding = FragmentWordsBinding.bind(view)
        with(binding) {
            requireCompatActivity().supportActionBar?.setTitle(R.string.words_title)

            wordsRecyclerView.setHasFixedSize(true)
            wordsRecyclerView.addItemDecoration(DividerItemDecoration(requireContext(), DividerItemDecoration.VERTICAL))
            wordsRecyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
                override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                    super.onScrolled(recyclerView, dx, dy)
                    if (wordActionMode == null) {
                        if (dy > 0) {
                            addWordFab.hide()
                        } else {
                            addWordFab.show()
                        }
                    }
                }
            })
            wordsRecyclerView.adapter = wordAdapter

            addWordFab.setOnClickListener { showAddEditWordDialog(null) }
        }

        viewModel.viewState.observe(viewLifecycleOwner, { renderViewState(it) })
    }

    override fun onDestroyView() {
        binding.wordsRecyclerView.adapter = null
        _binding = null
        super.onDestroyView()
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.words, menu)
        val searchItem = menu.findItem(R.id.menu_search_words)
        val searchView = searchItem.actionView as SearchView
        val searchManager = requireActivity().getSystemService(Context.SEARCH_SERVICE) as SearchManager
        searchView.setSearchableInfo(
            searchManager.getSearchableInfo(ComponentName(requireActivity(), MainActivity::class.java))
        )
        searchView.imeOptions = EditorInfo.IME_ACTION_SEARCH
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean = false

            override fun onQueryTextChange(newText: String): Boolean {
                viewModel.onSearchWordChanged(newText)
                return true
            }
        })
        searchItem.setOnActionExpandListener(object : MenuItem.OnActionExpandListener {
            override fun onMenuItemActionExpand(item: MenuItem): Boolean = true

            override fun onMenuItemActionCollapse(item: MenuItem): Boolean {
                viewModel.onCloseSearchWordClicked()
                return true
            }
        })
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.menu_sort_words) {
            showSortWordsDialog()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun renderViewState(viewState: WordsViewModel.ViewState) {
        showEmptyProgress(viewState.emptyProgress)
        showEmptyData(viewState.emptyData)
        showEmptyError(viewState.emptyError.first, viewState.emptyError.second)
        showEmptySearchResult(viewState.emptySearchResult.first, viewState.emptySearchResult.second)
        showWords(viewState.words.first, viewState.words.second)
    }

    private fun showEmptyProgress(show: Boolean) {
        binding.emptyProgressBar.isVisible = show
    }

    private fun showEmptyData(show: Boolean) {
        binding.emptyDataTextView.isVisible = show
    }

    private fun showEmptyError(show: Boolean, message: String?) {
        binding.errorTextView.text = message
        binding.errorTextView.isVisible = show
    }

    private fun showEmptySearchResult(show: Boolean, query: String?) {
        query?.let {
            val emptyResult = String.format(getString(R.string.words_empty_search_result), it.htmlEncode())
            binding.emptySearchResultTextView.text = emptyResult.parseAsHtml()
        }
        binding.emptySearchResultTextView.isVisible = show
    }

    private fun showWords(show: Boolean, words: List<Word>) {
        wordAdapter.submitList(words)
        binding.wordsRecyclerView.isVisible = show
    }

    private fun toggleSelection(position: Int) {
        wordAdapter.toggleSelection(position)
        val count = wordAdapter.getSelectedWordCount()
        if (count == 0) {
            wordActionMode?.finish()
        } else {
            wordActionMode?.title = count.toString()
            wordActionMode?.invalidate()
        }
    }

    private fun showAddEditWordDialog(word: Word?) {
        AddEditWordDialog.newInstance(word?.toUiModel())
            .show(parentFragmentManager, null)
    }

    private fun showSortWordsDialog() {
        setFragmentResultListener(SortWordsDialog.REQUEST_SORT_MODE) { _, bundle ->
            val sortMode = bundle.getSerializable(SortWordsDialog.RESULT_SORT_MODE) as Word.SortMode
            viewModel.onSortWordSelected(sortMode)
        }
        SortWordsDialog.newInstance(viewModel.getWordSortMode())
            .show(requireActivity().supportFragmentManager, null)
    }

    private fun showSelectWordCategoriesDialog() {
        setFragmentResultListener(SelectWordCategoryDialog.REQUEST_SELECT_WORD_CATEGORY) { _, bundle ->
            val category = bundle.getString(SelectWordCategoryDialog.RESULT_SELECT_WORD_CATEGORY)!!
            val words = wordAdapter.getSelectedWordPositions()
                .map { wordAdapter.currentList[it] }
            viewModel.onSelectWordCatalogClicked(words.toMutableList(), category)
            wordActionMode?.finish()
            showSnackbar(R.string.words_success_move)
        }
        SelectWordCategoryDialog.newInstance(R.string.words_action_move)
            .show(requireActivity().supportFragmentManager, null)
    }

    private inner class WordActionModeCallback : ActionMode.Callback {

        private lateinit var editWordItem: MenuItem

        override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
            mode.menuInflater.inflate(R.menu.selected_word, menu)
            editWordItem = menu.findItem(R.id.menu_edit_word)
            binding.addWordFab.hide()
            return true
        }

        override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
            editWordItem.isVisible = wordAdapter.getSelectedWordCount() == 1
            return true
        }

        override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean =
            when (item.itemId) {
                R.id.menu_move_word -> {
                    showSelectWordCategoriesDialog()
                    true
                }
                R.id.menu_edit_word -> {
                    val firstPos = wordAdapter.getSelectedWordPositions().first()
                    val word = wordAdapter.currentList[firstPos]
                    showAddEditWordDialog(word)
                    mode.finish()
                    true
                }
                R.id.menu_delete_word -> {
                    val words = wordAdapter.getSelectedWordPositions()
                        .map { wordAdapter.currentList[it] }
                    viewModel.onDeleteWordsClicked(words)
                    mode.finish()
                    true
                }
                else -> false
            }

        override fun onDestroyActionMode(mode: ActionMode) {
            wordAdapter.clearSelection()
            wordActionMode = null
            binding.addWordFab.show()
        }
    }
}
